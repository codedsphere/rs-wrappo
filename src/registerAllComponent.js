import map from 'lodash-es/map';

const registerComponent = components => tag => {
  const TagComponent = components[tag];
  if (TagComponent) {
    return { Tag: TagComponent, Type: 'Component' };
  }
  return { Tag: null, Type: '' };
};

function getCurrentTag(getTag, tagName) {
  const { Tag, Type } = getTag(tagName);
  if (Tag) {
    return { Tag, Type };
  }

  return null;
}

export default (TagMainOut, TagOverrideOut) => Tags => {
  const TagMain = registerComponent(TagMainOut);
  let TagOverride;
  if (TagOverrideOut) {
    TagOverride = registerComponent(TagOverrideOut);
  }

  const components = {};
  map(Tags, tag => {
    if (TagOverride) {
      const CurrentTag = getCurrentTag(TagOverride, tag);
      components[tag] = CurrentTag || getCurrentTag(TagMain, tag);
    }
    else {
      components[tag] = getCurrentTag(TagMain, tag);
    }
  });
  return { ...components };
};
