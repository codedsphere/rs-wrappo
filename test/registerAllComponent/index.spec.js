import React from 'react';
import { shallow } from 'enzyme';
import registerAllComponent from 'registerAllComponent';
import main from '../__mocks__/client/main';
import override from '../__mocks__/client/override';


describe('registerAllComponent', () => {
  it('should have place only main value', () => {
    const MainComponentRendered = registerAllComponent(main);

    const { assets } = MainComponentRendered(['assets']);
    const { main: Main } = assets.Tag;

    expect(assets).toMatchSnapshot();
    expect(shallow(<Main />)).toMatchSnapshot();
    expect(shallow(<Main />).contains(<div>Test Main</div>)).toBe(true);
  });
});

describe('registerAllComponent', () => {
  it('negative test of only main value', () => {
    const MainComponentRendered = registerAllComponent({});

    const { assets } = MainComponentRendered(['assets']);

    expect(assets).toBe(null);
  });
});

describe('registerAllComponent', () => {
  it('should have place a correct override value', () => {
    const MainComponentRendered = registerAllComponent(main, override);

    const { assets } = MainComponentRendered(['assets']);
    const { main: Main } = assets.Tag;

    expect(assets).toMatchSnapshot();
    expect(shallow(<Main />)).toMatchSnapshot();
    expect(shallow(<Main />).contains(<div>Test Override</div>)).toBe(true);
  });
});
