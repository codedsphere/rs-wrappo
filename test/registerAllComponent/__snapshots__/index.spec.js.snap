// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerAllComponent should have place a correct override value 1`] = `
Object {
  "Tag": Object {
    "main": [Function],
  },
  "Type": "Component",
}
`;

exports[`registerAllComponent should have place a correct override value 2`] = `
<div>
  Test Override
</div>
`;

exports[`registerAllComponent should have place only main value 1`] = `
Object {
  "Tag": Object {
    "main": [Function],
  },
  "Type": "Component",
}
`;

exports[`registerAllComponent should have place only main value 2`] = `
<div>
  Test Main
</div>
`;
